File: src/test/java/com/raychenon/leetcode/MaximalSquareTest.java
package com.raychenon.leetcode;

import org.junit.BeforeClass;
import org.junit.Test;

import static junit.framework.TestCase.assertEquals;

/**
 * User: raychenon
 * Date: 22/5/2020
 * https://leetcode.com/problems/maximal-square/
 */
public class MaximalSquareTest {

    private static MaximalSquare maximalSquare;

    @BeforeClass
    public static void setUp() {
        maximalSquare = new MaximalSquare();
    }

    @Test
    public void emptyMatrixTest() {
        char[][] matrix = {{}};
        evaluateMaximalSquareOriginal(0, matrix);
    }

    @Test
    public void example1Test() {

        char[][] matrix = {
                {'1', '0', '1', '0', '0'},
                {'1', '0', '1', '1', '1'},
                {'1', '1', '1', '1', '1'},
                {'1', '0', '0', '1', '0'}
        };

        evaluateMaximalSquareOriginal(4, matrix);
    }

    @Test
    public void biggerMatrixTest() {

        // 2 matrix of size 9
        char[][] matrix = {
                {'1', '1', '0', '0', '0'},
                {'1', '1', '1', '1', '0'},
                {'1', '1', '1', '1', '1'},
                {'1', '1', '1', '1', '0'}
        };

        assertEquals(false, maximalSquare.isSquare(matrix,1,3,3,5));
        assertEquals(true, maximalSquare.isSquare(matrix,1,3,1,3));
        evaluateMaximalSquareOriginal(9, matrix);
    }

    @Test
    public void MatrixCoversRightColsTest() {

        // 2 matrix of size 4
        char[][] matrix = {
                {'1', '1', '0', '0'},
                {'1', '0', '1', '1'},
                {'1', '1', '1', '1'},
                {'1', '0', '1', '1'}
        };

        evaluateMaximalSquareOriginal(4, matrix);
    }

    @Test
    public void MatrixThatBroke1DTest() {
        char[][] matrix = {
                {'1', '0', '1', '1', '1'},
                {'0', '1', '0', '1', '0'},
                {'1', '1', '0', '1', '1'},
                {'1', '1', '0', '1', '1'},
                {'0', '1', '1', '1', '1'}
        };

        evaluateMaximalSquareOriginal(4, matrix);
    }

    void evaluateMaximalSquareOriginal(int expected, char[][] matrix) {
        assertEquals(expected, maximalSquare.maximalSquareOriginal(matrix));
        assertEquals(expected, maximalSquare.maximalSquareDP(matrix));
        assertEquals(expected, maximalSquare.maximalSquareDP1Dmemory(matrix));
    }


}

File: src/test/kotlin/com/raychenon/kotlin/leetcode/MaximalSquareTest.kt
package com.raychenon.kotlin.leetcode

import junit.framework.TestCase
import org.junit.Test

/**
 * User: raychenon
 * Date: 23/5/2020
 * https://leetcode.com/problems/maximal-square/
 */
class MaximalSquareTest {

    @Test
    fun emptyMatrixTest() {
        val matrix = arrayOf(CharArray(0))
        evaluateMaximalSquareOriginal(0, matrix)
    }
    
    @Test
    fun example1Test() {
        val matrix = arrayOf(
            charArrayOf('1', '0', '1', '0', '0'),
            charArrayOf('1', '0', '1', '1', '1'),
            charArrayOf('1', '1', '1', '1', '1'),
            charArrayOf('1', '0', '0', '1', '0')
        )
        evaluateMaximalSquareOriginal(4, matrix)
    }

    @Test
    fun biggerMatrixTest() {

        // 2 matrix of size 9
        val matrix = arrayOf(
            charArrayOf('1', '1', '0', '0', '0'),
            charArrayOf('1', '1', '1', '1', '0'),
            charArrayOf('1', '1', '1', '1', '1'),
            charArrayOf('1', '1', '1', '1', '0')
        )
        evaluateMaximalSquareOriginal(9, matrix)
    }

    @Test
    fun MatrixThatBroke1DTest() {
        val matrix = arrayOf(
            charArrayOf('1', '0', '1', '1', '1'),
            charArrayOf('0', '1', '0', '1', '0'),
            charArrayOf('1', '1', '0', '1', '1'),
            charArrayOf('1', '1', '0', '1', '1'),
            charArrayOf('0', '1', '1', '1', '1')
        )
        evaluateMaximalSquareOriginal(4, matrix)
    }

    fun evaluateMaximalSquareOriginal(expected: Int, matrix: Array<CharArray>) {
        TestCase.assertEquals(expected, MaximalSquare.maximalSquareDP2Darray(matrix))
        TestCase.assertEquals(expected, MaximalSquare.maximalSquareDP1Darray(matrix))
    }
}